src/coreclr/jit/morph.cpp

```diff
@@ -6287,14 +6299,18 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
                 tree->SetOper(GT_IND);
                 tree->AsOp()->gtOp1 = addr;
 
-                if (isStaticReadOnlyInited)
+                if (isBoxedStatic)
+                {
+                    // The box for the static cannot be null, and is logically invariant, since it
+                    // represents (a base for) the static's address.
+                    tree->gtFlags |= (GTF_IND_INVARIANT | GTF_IND_NONFAULTING | GTF_IND_NONNULL);
+                }
+                else if (isStaticReadOnlyInited)
                 {
                     JITDUMP("Marking initialized static read-only field '%s' as invariant.\n", eeGetFieldName(symHnd));
 
                     // Static readonly field is not null at this point (see getStaticFieldCurrentClass impl).
                     tree->gtFlags |= (GTF_IND_INVARIANT | GTF_IND_NONFAULTING | GTF_IND_NONNULL);
-                    tree->gtFlags &= ~GTF_ICON_INITCLASS;
```

Comment:
can you put an assert that INITCLASS is never set here?
for `static readonly field` change I used `jit-utils --ctors` mode becuase SPMI won't catch it

Reply:
I could, but it doesn't seem very necessary? The `GTF_ICON_INITCLASS` was cleared on `tree`, which didn't have it set in the first place, because it is the `IND`, not the constant node (the value for this flag overlaps with `GTF_IND_UNALIGNED` but we could never have meaningful `GTF_IND_UNALIGNED` here because this path is only reached for `ref`-typed fields, which are always properly aligned).

Then the actual `GTF_ICON_INITCLASS` was implicitly cleared when we did `addr->gtFlags = GTF_ICON_CONST_PTR`, now it is just not set for the `isStaticReadOnlyInited` case. So all in all the behavior should be exactly the same as before.

Reply:
ok makes sense